pkg/server/testing/fake_cni_plugin.go

```diff
@@ -0,0 +1,187 @@
+	ContainerID string
+}
+
+// FakeCNIPlugin is a fake plugin used for test.
+type FakeCNIPlugin struct {
+	sync.Mutex
+	called []CalledDetail
+	errors map[string]error
+	IPMap  map[CNIPluginArgument]string
+}
+
+// getError get error for call
+func (f *FakeCNIPlugin) getError(op string) error {
+	err, ok := f.errors[op]
+	if ok {
+		delete(f.errors, op)
+		return err
+	}
+	return nil
+}
+
+// InjectError inject error for call
+func (f *FakeCNIPlugin) InjectError(fn string, err error) {
+	f.Lock()
+	defer f.Unlock()
+	f.errors[fn] = err
+}
+
+// InjectErrors inject errors for calls
+func (f *FakeCNIPlugin) InjectErrors(errs map[string]error) {
+	f.Lock()
+	defer f.Unlock()
+	for fn, err := range errs {
+		f.errors[fn] = err
+	}
+}
+
+// ClearErrors clear errors for call
+func (f *FakeCNIPlugin) ClearErrors() {
+	f.Lock()
+	defer f.Unlock()
+	f.errors = make(map[string]error)
+}
+
+func (f *FakeCNIPlugin) appendCalled(name string, argument interface{}) {
+	call := CalledDetail{Name: name, Argument: argument}
+	f.called = append(f.called, call)
+}
+
+// GetCalledNames get names of call
+func (f *FakeCNIPlugin) GetCalledNames() []string {
+	f.Lock()
+	defer f.Unlock()
+	names := []string{}
+	for _, detail := range f.called {
+		names = append(names, detail.Name)
+	}
+	return names
+}
+
+// GetCalledDetails get detail of each call.
+func (f *FakeCNIPlugin) GetCalledDetails() []CalledDetail {
+	f.Lock()
+	defer f.Unlock()
+	// Copy the list and return.
+	return append([]CalledDetail{}, f.called...)
+}
+
+// SetFakePodNetwork sets the given IP for given arguments.
+func (f *FakeCNIPlugin) SetFakePodNetwork(netnsPath string, namespace string, name string, containerID string, ip string) {
+	arg := CNIPluginArgument{netnsPath, namespace, name, containerID}
+	f.IPMap[arg] = ip
+}
+
+// NewFakeCNIPlugin create a FakeCNIPlugin.
+func NewFakeCNIPlugin() ocicni.CNIPlugin {
+	return &FakeCNIPlugin{
+		errors: make(map[string]error),
+		IPMap:  make(map[CNIPluginArgument]string),
+	}
+}
+
+// Name return the name of fake CNI plugin.
+func (f *FakeCNIPlugin) Name() string {
+	return "fake-CNI-plugin"
+}
+
+// SetUpPod setup the network of PodSandbox.
+func (f *FakeCNIPlugin) SetUpPod(netnsPath string, namespace string, name string, containerID string) error {
+	f.Lock()
+	defer f.Unlock()
+
+	arg := CNIPluginArgument{netnsPath, namespace, name, containerID}
+	f.appendCalled("SetUpPod", arg)
+
+	if err := f.getError("SetUpPod"); err != nil {
+		return err
+	}
+
+	f.IPMap[arg] = generateIP()
+
+	return nil
+}
+
+// TearDownPod teardown the network of PodSandbox.
+func (f *FakeCNIPlugin) TearDownPod(netnsPath string, namespace string, name string, containerID string) error {
+	f.Lock()
+	defer f.Unlock()
+
+	arg := CNIPluginArgument{netnsPath, namespace, name, containerID}
+	f.appendCalled("TearDownPod", arg)
+
+	if err := f.getError("TearDownPod"); err != nil {
+		return err
+	}
+
+	_, ok := f.IPMap[arg]
+	if !ok {
+		return fmt.Errorf("failed to find the IP")
```

Comment:
I'm not sure whether we should return error here. CNI doesn't define this, some plugin returns error in this case, some not. Let's treat it as it will return an error then.